### PR TITLE
UI for creating builds + connection to backend

### PR DESCRIPTION
**Pull request by wojta**
There is PR for the creating build configs.
https://issues.jboss.org/browse/AEROGEAR-7926

### Parts
- [x]  UI for creating builds
- [x]  Gathering state from UI
- [x] Calling backend
- [x] Validation
- [x] Entering SSH key and Android keystore as File Upload
- [x] Making UI more 1resemble flats
- [x] Attach edit and create new build config from build configs overview screen
- [x] Rework to use Redux everywhere

### CODE REVIEW

File: ui/package.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "babel-eslint": "^9.0.0",
```
Comment (jhellar) [inline ui/package.json]: Is this needed? `babel-eslint` should be already there as a dependency of `react-scripts` and `eslint-plugin-patternfly-react`.

File: ui/src/DataService.js [hunk 2]
```diff
@@ -17,6 +17,7 @@ if (process.env.NODE_ENV !== 'production') {
   baseUrl = 'http://localhost:8080/http://localhost:4000/api';
   wsUrl = 'ws://localhost:4000/api';
 }
+baseUrl = 'http://localhost:4000/api';
```
Comment (jhellar) [inline ui/src/DataService.js]: This should not be there IMO. The code above should handle `baseUrl` for both production (`make serve`) and debug (`./scripts/development.sh`).

File: ui/src/containers/Client.js [hunk 3]
```diff
@@ -56,6 +57,8 @@ class Client extends Component {
   }
 
   render() {
+    const { pathname } = this.props.location;
+    const clientId = pathname.substr(pathname.lastIndexOf('/') + 1);
```
Comment (jhellar) [inline ui/src/containers/Client.js]: You can get clientId from `this.props.match.params.id`

File: ui/src/components/build/CreateBuildConfigDialog.js [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+          <Row>
+            <Col lg={advancedOptions ? 8 : 12}>
+              <FormGroup>
+                <ControlLabel className="required">Git Repository URL</ControlLabel>
+                <FormControl type="text" onChange={e => this.sourceState.set('gitUrl', e.target.value)} value={gitUrl} />
+                <HelpBlock>Git URL of the source code to build.</HelpBlock>
+                <HelpBlock>
+            View the
+                  {' '}
+                  <a href="#" onClick={() => this.setState({ advancedOptions: true })}>advanced options</a>
```
Comment (jhellar) [inline ui/src/components/build/CreateBuildConfigDialog.js]: In OpenShift when you click the `advanced options`, it hides. Maybe we should do the same here.

**Comment by jhellar** [comment]
Few more comments:

- package-lock.json should be updated (should be generated with npm from nodejs 8)
- inputs from the create build config modal should be reset to their default values when the modal is closed
- `NoBuildConfig` component - components that access redux store or dispatch actions are now stored in `containers` folder. not sure if this is the best approach, but better to be consistent

File: ui/package.json (hunk 1 above)
Comment (wojta) [reply ui/package.json]: probably not

File: ui/src/components/build/CreateBuildConfigDialog.js (hunk 4 above)
Comment (wojta) [reply ui/src/components/build/CreateBuildConfigDialog.js]: ok

File: ui/src/components/build/CreateBuildConfigDialog.js [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    update: PropTypes.bool,
+    buildConfig: PropTypes.object,
+  };
+
+  constructor(props) {
+    super(props);
+    const { buildConfig } = props;
+    this.state = {
+      config: {
+        source: { authType: BUILD_AUTH_TYPE_PUBLIC, gitRef: 'master', jenkinsFilePath: '/' },
```
[the remark this answers is not in the document]
Comment (wojta) [reply ui/src/components/build/CreateBuildConfigDialog.js]: I was wondering about this. Ok.

File: ui/src/containers/Client.js (hunk 3 above)
Comment (wojta) [reply ui/src/containers/Client.js]: thanks

File: ui/src/DataService.js (hunk 2 above)
Comment (wojta) [reply ui/src/DataService.js]: I left it there from debugging.

File: ui/src/components/build/CreateBuildConfigDialog.js [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+          <Grid fluid>
+            <Row>
+              <Col md={10}>
+                <div className="help-block">
+                  Client Build allows you to create a mobile client binary from a git source repository.
+                </div>
+                <Form>
+                  {this.renderNameSection()}
+                  {this.renderSourceSection()}
+                  {this.renderBuildSection()}
```
Comment (wei-lee) [inline ui/src/components/build/CreateBuildConfigDialog.js]: can we split those into smaller components? This file is too big and it also prevents those components to be reused.

File: ui/src/DataService.js [hunk 7]
```diff
@@ -132,15 +147,21 @@ const dataService = {
...
+      }
     });
     if (!response.ok) {
       throw Error(response.statusText);
     }
     const user = await response.json();
     return user;
-  }
+  },
+  logout: async () =>
```
Comment (wei-lee) [inline ui/src/DataService.js]: you might need a rebase. This function should be removed from master.

File: ui/src/components/build/CreateBuildConfigDialog.js [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_NAME,
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE,
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE_ALIAS,
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE_PASSWORD
+    ]);
+    this.iOSCredentialsState = new SubState(this, 'config.build.iosCredentials', iOSCredentialsValidation, [
+      KEY_CR_BUILD_IOS_CREDENTIALS,
+      KEY_CR_BUILD_IOS_CREDENTIALS_DEVELOPER_PROFILE,
+      KEY_CR_BUILD_IOS_CREDENTIALS_NAME
+    ]);
```
Comment (wei-lee) [inline ui/src/components/build/CreateBuildConfigDialog.js]: rather than defining the state of the sub components in the parent component, I am wondering if we should move those to the child components themselves? It feels like the parent component is knowing too much about the child components.

File: ui/src/components/build/create_build_config/BuildSection.js [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      {externalCredentials ? renderAndroidCredentials(component) : <React.Fragment />}
+      <FormGroup>
+        <ControlLabel>Environment Variables</ControlLabel>
+        <KeyValueEditor onChange={rows => onEnvVarsChange(configState, rows)} />
+      </FormGroup>
+    </div>
+  );
+}
+
+function renderAndroidCredentials(component) {
```
Comment (wei-lee) [inline ui/src/components/build/create_build_config/BuildSection.js]: I think this really should be a React component on its own.

File: ui/src/components/build/create_build_config/BuildSection.js [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+          >
+            Upload external credentials (Use this option if your Keystore is external to your source code)
+          </Checkbox>
+        </FormGroup>
+      ) : (
+        <React.Fragment />
+      )}
+      {platform === BUILD_PLATFORM_IOS ? renderIOSCredentials(component) : <React.Fragment />}
+      {externalCredentials ? renderAndroidCredentials(component) : <React.Fragment />}
+      <FormGroup>
```
Comment (wei-lee) [inline ui/src/components/build/create_build_config/BuildSection.js]: should this be its own component?

File: ui/src/components/build/create_build_config/BuildSection.js [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+          The entry name of the private key/certificate chain you want to use to sign your APK(s). This entry must exist
+          in the key store uploaded. If your key store contains only one key entry, which is the most common case, you
+          can leave this field blank.
+        </HelpBlock>
+      </FormGroup>
+    </div>
+  );
+}
+
+function renderIOSCredentials(component) {
```
Comment (wei-lee) [inline ui/src/components/build/create_build_config/BuildSection.js]: the same with this one, it should be a component of its own.

File: ui/src/components/build/create_build_config/NameSection.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import { KEY_CR_NAME } from '../Constants';
+import { FormGroup, FormControl, HelpBlock, ControlLabel } from 'patternfly-react';
+
+export function renderNameSection(configState) {
```
Comment (wei-lee) [inline ui/src/components/build/create_build_config/NameSection.js]: make it a react component.

File: ui/src/components/build/create_build_config/SourceSection.js [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        <HelpBlock>Upload your private SSH key file.</HelpBlock>
+        <FormControl
+          componentClass="textarea"
+          onChange={e => sshAuthState.set(KEY_CR_SSH_PRIVATE_KEY, e.target.value)}
+          value={privateKey}
+        />
+        <HelpBlock>Private SSH key file for Git authentication.</HelpBlock>
+      </FormGroup>
+    </React.Fragment>
+  );
```
Comment (wei-lee) [inline ui/src/components/build/create_build_config/SourceSection.js]: I think these should really be react components

File: ui/src/components/build/create_build_config/Validations.js [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    case KEY_CR_BUILD_ANDROID_CREDENTIALS_NAME:
+    case KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE:
+    case KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE_ALIAS:
+    case KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE_PASSWORD:
+      return validateNotEmpty(value);
+    default:
+      return VALIDATION_OK;
+  }
+}
+export function iOSCredentialsValidation(values, key, value) {
```
Comment (wei-lee) [inline ui/src/components/build/create_build_config/Validations.js]: those validation methods should be inside the corresponding components.

File: ui/src/components/common/SubState.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
+import { set as _set, get as _get } from 'lodash-es';
+import { VALIDATION_WARN, VALIDATION_OK } from './Validation';
+
+export class SubState {
```
Comment (wei-lee) [inline ui/src/components/common/SubState.js]: do we really need this class if each components manage its own state?

**Comment by wei-lee** [comment]
Also I think we should really use Redux in some of the components here to help manage the state between the parent and child components.

File: ui/src/components/build/Constants.js [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+/* BuildConfig platform */
+export const BUILD_PLATFORM_ANDROID = 'android';
+export const BUILD_PLATFORM_IOS = 'iOS';
+
+/* BuildConfig buildType */
+export const BUILD_TYPE_DEBUG = 'debug';
+export const BUILD_TYPE_RELEASE = 'release';
+
+export const KEY_HIDE_PLATDORM = 'hidePlatform';
```
Comment (jhellar) [inline ui/src/components/build/Constants.js]: typo `PLATDORM`

File: ui/src/components/build/CreateBuildConfigDialog.js (hunk 8 above)
Comment (jhellar) [reply ui/src/components/build/CreateBuildConfigDialog.js]: IMO this top-level component (`CreateBuildConfigDialog`) needs to store all this information in it's state, so that when user clicks `Create` button, it has all the information needed to send the request.

Also relevant parts of this state should be passed as props to the descendant components and when some field changes, the child component should call callback of this top-level component so that it can update it's state.

Validation should be done in child components IMO.

File: ui/src/components/build/CreateBuildConfigDialog.js [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+      validVal = validVal && this.basicAuthState.isAllValid();
+    if (this.sourceState.get(KEY_CR_SOURCE_AUTH_TYPE) === BUILD_AUTH_TYPE_SSH)
+      validVal = validVal && this.sshAuthState.isAllValid();
+    if (this.state.valid !== validVal) this.setState({ valid: validVal });
+  }
+
+  onSaveBuildConfig = () => {
+    const { config } = this.state;
+    const { onSave, update = false, initialConfig } = this.props;
+    onSave && onSave({ update, config: { ...initialConfig, ...config } });
```
Comment (jhellar) [inline ui/src/components/build/CreateBuildConfigDialog.js]: Is `{ ...initialConfig, ...config }` needed? Should not `this.state.config` be initializes with `this.props.config`/`this.props.initialConfig` in constructor? What is difference between `this.props.config` and `this.props.initialConfig`?

File: ui/src/components/build/CreateBuildConfigDialog.js [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+      KEY_CR_BUILD_IOS_CREDENTIALS_DEVELOPER_PROFILE,
+      KEY_CR_BUILD_IOS_CREDENTIALS_NAME
+    ]);
+  }
+
+  componentDidUpdate(prevProps) {
+    this.validate();
+    const { initialConfig: prevConfig = {} } = prevProps;
+    const { initialConfig = {} } = this.props;
+    if (!isEqual(prevConfig, initialConfig)) {
```
Comment (jhellar) [inline ui/src/components/build/CreateBuildConfigDialog.js]: Why do we need to to update state based on props? IMO `CreateBuildConfigDialog` should only be initializes with config (in case of buildConfig update) in constructor of this component.

File: ui/src/components/build/CreateBuildConfigDialog.js [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+          <Grid fluid>
+            <Row>
+              <Col md={12}>
+                <div className="help-block">
+                  Client Build allows you to create a mobile client binary from a git source repository.
+                </div>
+                <Form>
+                  {renderNameSection(this.configState)}
+                  {renderSourceSection(this)}
+                  {renderBuildSection(this)}
```
Comment (jhellar) [inline ui/src/components/build/CreateBuildConfigDialog.js]: IMO only parameters that should be passed to these render methods are relevant parts of the state and callbacks for handling changes.

File: ui/src/components/build/CreateBuildConfigDialog.js [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+    </div>
+  );
+
+  getValidationState = () => {};
+
+  render() {
+    const { show, title, onCancel, update } = this.props;
+    const { valid = false } = this.state;
+    return (
+      <Modal show={show} container={this}>
```
Comment (jhellar) [inline ui/src/components/build/CreateBuildConfigDialog.js]: If we use this `container` property should we also add these [1] css styles?

[1] https://react-bootstrap.github.io/components/modal/#modals-contained

File: ui/src/components/build/MobileClientBuildOverviewList.js [hunk 21]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { ListView } from 'patternfly-react';
+import MobileClientBuildListItem from './MobileClientBuildListItem';
+import NoBuildConfig from './NoBuildConfig';
+
+export const MobileClientBuildOverviewList = ({ buildConfigs, appName, config }) => (
+  <div>
+    {// TODO: Mobile builds are not avaialble for Xamarin.
+    buildConfigs.length === 0 ? (
+      <NoBuildConfig config={config} />
```
Comment (jhellar) [inline ui/src/components/build/MobileClientBuildOverviewList.js]: IMO we don't need to pass `config` here. `CreateBuildConfigDialog` component can determine `clientType` and `clientId` by accessing redux store directly.

File: ui/src/components/build/NoBuildConfig.js [hunk 22]
```diff
@@ -1,16 +1,35 @@
...
         <h2>No Build Config</h2>
         <p>Create a mobile build config to create a mobile client build.</p>
-        <Button bsStyle="primary" bsSize="large">Create Build</Button>
+        <Button bsStyle="primary" bsSize="large" onClick={() => this.setState({ show: true })}>
+          Create build config
+        </Button>
+        {
+          <BuildConfigDialog
+            update={false}
+            initialConfig={config}
```
Comment (jhellar) [inline ui/src/components/build/NoBuildConfig.js]: Same here - IMO this is not needed, better to get the values from redux store.

File: ui/src/containers/BuildConfigDialog.js [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+        onSave={this.buildUpdated}
+        onCancel={this.close}
+      />
+    );
+  };
+}
+
+function mapStateToProps(state) {
+  return {
+    config: state
```
Comment (jhellar) [inline ui/src/containers/BuildConfigDialog.js]: Why is this needed? I don't see it used anywhere.

File: ui/src/containers/BuildConfigDialog.js [hunk 24]
```diff
@@ -0,0 +1,53 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import CreateBuildConfigDialog from '../components/build/CreateBuildConfigDialog';
+import { createBuildConfig } from '../actions/buildConfigs';
+
+class BuildConfigDialog extends Component {
+  buildUpdated = configUpdate => {
+    const config = Object.assign({}, configUpdate.config);
+    this.props.createBuildConfig(config);
```
Comment (jhellar) [inline ui/src/containers/BuildConfigDialog.js]: BuildConfig update is not yet implemented.

File: ui/src/components/build/create_build_config/BuildSection.js [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+  configState.set(
+    KEY_CR_ENV_VARS,
+    rows.map(({ key, value }) => ({
+      name: key,
+      value
+    }))
+  );
+}
+
+export function renderBuildSection(component) {
```
Comment (jhellar) [inline ui/src/components/build/create_build_config/BuildSection.js]: I also think that this should be component.

File: ui/src/components/build/create_build_config/SourceSection.js (hunk 13 above)
Comment (jhellar) [reply ui/src/components/build/create_build_config/SourceSection.js]: agree with @wei-lee

File: ui/src/components/build/create_build_config/Validations.js (hunk 14 above)
Comment (jhellar) [reply ui/src/components/build/create_build_config/Validations.js]: also agree

File: ui/src/components/common/UploadControl.js [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+class UploadControl extends Component {
+  localUpload = e => {
+    const { files } = e.target;
+    if (!files.length) {
+      return;
+    }
+
+    const file = files[0];
+    const start = 0;
+    const stop = file.size - 1;
```
Comment (jhellar) [inline ui/src/components/common/UploadControl.js]: why do we deduct 1 here, if we add it later in the code?

File: ui/src/components/build/CreateBuildConfigDialog.js (hunk 20 above)
Comment (wojta) [reply ui/src/components/build/CreateBuildConfigDialog.js]: ok

File: ui/src/components/build/CreateBuildConfigDialog.js [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+      valid: false,
+      [KEY_HIDE_PLATDORM]: false,
+      config: {
+        [KEY_CR_SOURCE]: {
+          [KEY_CR_SOURCE_AUTH_TYPE]: BUILD_AUTH_TYPE_PUBLIC,
+          [KEY_CR_SOURCE_GITREF]: 'master',
+          [KEY_CR_SOURCE_JENKINS_FILE_PATH]: './'
+        },
+        [KEY_CR_BUILD]: { [KEY_CR_BUILD_PLATFORM]: BUILD_PLATFORM_ANDROID, [KEY_CR_BUILD_TYPE]: BUILD_TYPE_DEBUG },
+        ...this.props.config
```
[the remark this answers is not in the document]
Comment (wojta) [reply ui/src/components/build/CreateBuildConfigDialog.js]: `propTypes` gets removed, at least for now

File: ui/src/components/build/CreateBuildConfigDialog.js (hunk 8 above)
Comment (wojta) [reply ui/src/components/build/CreateBuildConfigDialog.js]: Reduxed...

File: ui/src/components/build/CreateBuildConfigDialog.js (hunk 18 above)
Comment (wojta) [reply ui/src/components/build/CreateBuildConfigDialog.js]: removed

File: ui/src/components/build/CreateBuildConfigDialog.js (hunk 17 above)
Comment (wojta) [reply ui/src/components/build/CreateBuildConfigDialog.js]: removed

File: ui/src/components/build/create_build_config/BuildSection.js (hunk 10 above)
Comment (wojta) [reply ui/src/components/build/create_build_config/BuildSection.js]: changed to its own component

File: ui/src/containers/BuildConfigDialog.js (hunk 24 above)
Comment (wojta) [reply ui/src/containers/BuildConfigDialog.js]: renamed

File: ui/src/components/build/CreateBuildConfigDialog.js [hunk 28]
```diff
@@ -0,0 +1,267 @@
...
+                </FormGroup>
+              </Col>
+              ) : <React.Fragment />}
+          </Row>
+        </Grid>
+        {advancedOptions
+          ? (
+            <FormGroup>
+              <ControlLabel className="required">Jenkins file path</ControlLabel>
+              <FormControl type="text" defaultValue={jenkinsFilePath} onChange={e => this.sourceState.set('jenkinsFilePath', e.target.value)} value={jenkinsFilePath} />
```
[the remark this answers is not in the document]
Comment (wojta) [reply ui/src/components/build/CreateBuildConfigDialog.js]: ok

File: ui/src/components/build/CreateBuildConfigDialog.js (hunk 6 above)
Comment (wojta) [reply ui/src/components/build/CreateBuildConfigDialog.js]: splitted

File: ui/src/components/build/create_build_config/BuildSection.js (hunk 11 above)
Comment (wojta) [reply ui/src/components/build/create_build_config/BuildSection.js]: done

File: ui/src/components/build/create_build_config/Validations.js (hunk 14 above)
Comment (wojta) [reply ui/src/components/build/create_build_config/Validations.js]: Not sure about this. It's validation of the state, rather than with components.

File: ui/src/components/common/SubState.js (hunk 15 above)
Comment (wojta) [reply ui/src/components/common/SubState.js]: removed file

File: ui/src/components/build/CreateBuildConfigDialog.js (hunk 19 above)
Comment (wojta) [reply ui/src/components/build/CreateBuildConfigDialog.js]: Redux solves this now.

File: ui/src/components/build/MobileClientBuildOverviewList.js (hunk 21 above)
Comment (wojta) [reply ui/src/components/build/MobileClientBuildOverviewList.js]: passing `clientInfo` now, in redux store it's not available specifically what client app is selected, it's determined in `Client.js`

File: ui/src/components/build/NoBuildConfig.js (hunk 22 above)
Comment (wojta) [reply ui/src/components/build/NoBuildConfig.js]: same as above

File: ui/src/containers/BuildConfigDialog.js (hunk 23 above)
Comment (wojta) [reply ui/src/containers/BuildConfigDialog.js]: removed

File: ui/src/components/common/UploadControl.js (hunk 26 above)
Comment (wojta) [reply ui/src/components/common/UploadControl.js]: fixed

**Comment by wojta** [comment]
@wei-lee, @jhellar Please can you review it again. 
I still like to add some tests of that components, it's pretty complex.

File: ui/src/components/build/create_build_config/BuildSection.js (hunk 9 above)
Comment (wojta) [reply ui/src/components/build/create_build_config/BuildSection.js]: made own component

File: ui/package.json [hunk 29]
```diff
@@ -32,16 +33,16 @@
     "start": "npm-run-all -p watch-css start-js",
     "build-js": "react-scripts build",
     "build": "npm-run-all build-css build-js",
-    "test": "react-scripts test --env=jsdom",
-    "coverage": "react-scripts test --env=jsdom --coverage",
+    "test": "react-scripts test --env=jsdom --transformIgnorePatterns \"node_modules/(?!lodash-es)/\"",
+    "coverage": "react-scripts test --env=jsdom --coverage --transformIgnorePatterns \"node_modules/(?!lodash-es)/\"",
```
Comment (wei-lee) [inline ui/package.json]: @jhellar I think you were having some problems with using `lodash` in the tests before, right? Does this solve your problem?

File: ui/src/DataService.js [hunk 30]
```diff
@@ -78,16 +78,32 @@ const dataService = {
...
+  createBuildConfig: async config => {
+    const response = await fetch(`${baseUrl}/buildconfigs`, {
+      method: 'POST',
+      cache: 'no-cache',
+      credentials: 'same-origin',
+      headers: {
+        'Content-Type': 'application/json; charset=utf-8'
       },
-      body: JSON.stringify(app),
+      body: JSON.stringify(config)
```
Comment (wei-lee) [inline ui/src/DataService.js]: we can use the `requestConfig` method above.

File: ui/package.json (hunk 29 above)
Comment (jhellar) [reply ui/package.json]: @wei-lee yes, I think so. there was just one place where I used lodash, so I replaced it with plain  js, but good to have this

Comment (wojta) [reply ui/package.json]: Is it bad to use `lodash`? I really like to use the get, set and other useful methods, that are pretty messy in plain Javascript.

Comment (wei-lee) [reply ui/package.json]: @wojta no, no problem at all to use in the test. It's just @jhellar was trying to do the same in the tests before and there was some issues with it. This helps solve his problem so I just ping him here.
